### PR TITLE
New bootstrap 0428

### PR DESCRIPTION
**Pull request by alivedise**

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b07dc54be6d7f4538ada1ad8e4dbaee25fbb6b31

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=230602f40be97f416f91a7dd80355f69d20abf1d

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=951c2e5d48f6629f1fee005187b689aa554960f5

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6b54cbee70206b2604c3d0d6612e30d641a034a9

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e8e4c2c49dc0af267028443d6d2881afc438fa8d

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=0ba3609c49c061c6730af04b65816401b4378bbd

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3f50091217931df6df9e328ed0888ea7e67c2bb7

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fc0bbbab513149cb6594ef59be926f38e97d4bea

File: apps/system/js/app.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    ],
+    start: function() {
+      window.performance.mark('loadEnd');
+      return LazyLoader.load(this.FILES).then(() => {
+        return this.bootstrap();
+      });
+    },
+    bootstrap: function() {
+      if (this._booted) {
+        return Promise.resolve();
```
Comment (timdream) [inline apps/system/js/app.js]: `throw new Error('App: bootstrap should not be called twice.');`

File: apps/system/js/base_module.js [hunk 2]
```diff
@@ -622,7 +655,7 @@
       this.switchLifeCycle('stopped');
     },
 
-    switchLifeCycle: function(state) {
+    switchLifeCycle: function(state, resolve, reject) {
```
Comment (timdream) [inline apps/system/js/base_module.js]: Where is `resolve, reject` coming from?

File: apps/system/js/bootstrap.js [hunk 3]
```diff
@@ -1,247 +1,24 @@
...
-    /** @global */
-    window.textSelectionDialog = new TextSelectionDialog();
-  }
-
-  function safelyLaunchFTU() {
-    window.addEventListener('homescreen-ready', function onHomescreenReady() {
-      window.removeEventListener('homescreen-ready', onHomescreenReady);
-      FtuLauncher.retrieve();
+(function() {
+  window.app = new App();
```
Comment (timdream) [inline apps/system/js/bootstrap.js]: No warning here on freezing the file.

File: apps/system/js/device_storage_watcher.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    }
+  };
+
+  // unit tests call init() manually
+  if (navigator.mozL10n) {
+    navigator.mozL10n.once(
+      DeviceStorageWatcher.init.bind(DeviceStorageWatcher));
+  }
+  exports.DeviceStorageWatcher = DeviceStorageWatcher;
+}(window));
```
Comment (timdream) [inline apps/system/js/device_storage_watcher.js]: new line.

You should turn on this settings in Sublime Text.

File: apps/system/lockscreen/js/lockscreen.js [hunk 5]
```diff
@@ -484,13 +487,29 @@
...
+      }.bind(this));
     }
   };
 
+  LockScreen.prototype.startConnectionInfoManager = function() {
+    LazyLoader.load(
+      ['shared/js/lockscreen_connection_info_manager.js']).then(() => {
+        this._lockscreenConnInfoManager =
+          new LockScreenConnInfoManager(this.connStates);
+      });
```
Comment (timdream) [inline apps/system/lockscreen/js/lockscreen.js]: `.catch`

File: apps/system/lockscreen/js/lockscreen_notifications.js [hunk 6]
```diff
@@ -18,9 +16,12 @@
...
     this.container = container;
-    this._lockScreenNotificationBuilder =
-      new LockScreenNotificationBuilder(this.container);
-    this._lockScreenNotificationBuilder.start();
+    LazyLoader.load(['lockscreen/js/lockscreen_notification_builder.js'])
+      .then(() => {
+        this._lockScreenNotificationBuilder =
+          new LockScreenNotificationBuilder(this.container);
+        this._lockScreenNotificationBuilder.start();
+      });
```
Comment (timdream) [inline apps/system/lockscreen/js/lockscreen_notifications.js]: .catch

File: apps/system/test/unit/app_test.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    fakeLauncherStartDeferred.resolve();
+    fakeCoreStartDeferred.resolve();
+    Promise.all([
+      fakeApplicationReadyDeferred.promise,
+      fakeLauncherStartDeferred.promise
+    ]).then(function() {
+      assert.isDefined(window.core);
+    });
+  });
+});
```
Comment (timdream) [inline apps/system/test/unit/app_test.js]: EOF

File: apps/system/test/unit/base_module_test.js [hunk 8]
```diff
@@ -35,6 +35,44 @@ suite('system/BaseModule', function() {
...
+  suite('Launching promise', function() {
+    test('start should resolve right away', function(done) {
+      var LaunchingPromiseTester = function() {};
+      BaseModule.create(LaunchingPromiseTester, {
+        name: 'LaunchingPromiseTester'
+      });
+      var lpt = BaseModule.instantiate('LaunchingPromiseTester');
+      lpt.start().then(function() {
+        done();
+      });
```
Comment (timdream) [inline apps/system/test/unit/base_module_test.js]: `.catch((e) => {throw e || 'Should not throw'; }).then(done, done);`

File: apps/system/test/unit/base_module_test.js [hunk 9]
```diff
@@ -35,6 +35,44 @@ suite('system/BaseModule', function() {
...
+        name: 'LaunchingPromiseTester',
+        _start: function() {
+          return new Promise(function(resolve) {
+            resolve();
+          });
+        }
+      });
+      var lpt = BaseModule.instantiate('LaunchingPromiseTester');
+      lpt.start().then(function() {
+        done();
```
Comment (timdream) [inline apps/system/test/unit/base_module_test.js]: .catch((e) => {throw e || 'Should not throw'; }).then(done, done);

File: apps/system/test/unit/bluetooth_core_test.js [hunk 10]
```diff
@@ -40,17 +44,17 @@ suite('system/BluetoothCore', function() {
...
 
-    test('read', function() {
-      assert.isTrue(MockLazyLoader.load.calledWith(
-        ['js/bluetooth.js']
-      ));
+    test('Start', function(done) {
+      subject.start().then(() => {
+        assert.isTrue(window.Bluetooth1.start.called);
+        done();
+      });
```
Comment (timdream) [inline apps/system/test/unit/bluetooth_core_test.js]: .catch((e) => {throw e || 'Should not throw'; }).then(done, done);

File: apps/system/js/base_module.js (hunk 2 above)
Comment (alivedise) [inline apps/system/js/base_module.js]: Removed

**Pull request closed without merging**